// ask-cli/lib/deploy/deploy-lambda.js
'use strict';

const createLambda = require('./lambda-operation/create-lambda');
const updateLambda = require('./lambda-operation/update-lambda');
const parser = require('../utils/skill-parser');
const profileHelper = require('../utils/profile-helper');
const fs = require('fs');
const installNodeModule = require('../utils/install-node-module');


module.exports.deployLambda = (skillId, skillInfo, profile, doDebug, callback) => {
    if (!skillId) {
        console.error('No skill ID found. Please first create skill, then deploy lambda.');
        process.exit(1);
    }

    let awsProfile = profileHelper.getAWSProfile(profile);
    if (!awsProfile) {
        console.warn('[Warn]: Lambda deployment skipped because AWS credentials for profile [' +
            profile + '] are missing. CLI lambda functionalities can be' +
            " enabled by running `ask init` again to add 'aws_profile' to ASK cli_config");
        if (typeof callback === 'function' && callback) {
            callback();
        }
        return;
    }
    let updateLambdaList = [];
    let createLambdaList = [];
    for (let domain of Object.keys(skillInfo.endpointsInfo)) {
        for (let region of Object.keys(skillInfo.endpointsInfo[domain])) {
            separateLambdaCreateUpdate(skillInfo, domain, region, updateLambdaList, createLambdaList, profile);
        }
    }
    if (updateLambdaList.length === 0 && createLambdaList.length === 0) {
        console.log('[Info]: No lambda functions need to be deployed.');
        callback();
        return;
    }

    installNodeModule.install(skillInfo, process.cwd(), doDebug, () => {
        createLambda.createLambda(skillId, skillInfo.skillName, createLambdaList, awsProfile, (generatedLambda) => {
            updateLambda.updateLambda(updateLambdaList, awsProfile, () => {
                console.log('Lambda deployment finished.');
                printLambdaDeployResults(createLambdaList,updateLambdaList);
                if (typeof callback === 'function' && callback) {
                    callback(generatedLambda);
                }
            });
        });
    });
};

function printLambdaDeployResults(createLambdaList, updateLambdaList){
    if(createLambdaList && createLambdaList.length > 0){
        console.log('Lambda function(s) created:');
        createLambdaList.forEach(metaData => {
            console.log('  [URI] ' + metaData.uri);
        });
    }
    if(updateLambdaList && updateLambdaList.length > 0){
        console.log('Lambda function(s) updated:');
        updateLambdaList.forEach(metaData => {
            console.log('  [URI] ' + metaData.uri);
        });
    }
}
/*
 * Separate Lambda by directly updating the createList amd updateList passed in as inputs
 *
 * Each list is consisted of metaData which contains following properties:
 * {uri, sourceDir, domain, region, customName, zipPath, arn}
 * 4 of them are guaranteed to exist: uri, sourceDir, domain, region
 *
 * For create deploy lambda, metaData has 4 at the beginning, and 2 extra after creation:
 * - customName: boolean, used to test if the createLambda's name is input by client
 *               or at 'ask new' stage (true)/ or just created on 'ask deploy'(false).
 * - arn: the actual lambda function arn from aws response.
 */
function separateLambdaCreateUpdate(skillInfo, domain, region, updateLambdaList, createLambdaList, profile) {
    let uri = skillInfo.endpointsInfo[domain][region].uri;
    let sourceDir = skillInfo.endpointsInfo[domain][region].sourceDir;

    if (!uri && sourceDir && sourceDir.length !== 0) {
        if (!fs.existsSync(sourceDir)) {
            console.error('Source directory is invalid, cannot deploy lambda function(s)');
            process.exit(1);
        }
        createLambdaList.push({
            uri: 'ask-' + domain + '-' + skillInfo.skillName + '-' + profile,
            sourceDir: sourceDir,
            domain: domain,
            region: region,
            customName: false
        });
    }
    if (parser.isAbsoluteURL(uri) && uri.substr(0, 6) === 'https:') {
        return;
    }
    if (parser.isAbsoluteURL(uri) && uri.substr(0, 4) === 'arn:') {
        if (!sourceDir) {
            return;
        } else if (!fs.existsSync(sourceDir)) {
            console.error('Source directory is invalid, cannot deploy lambda function(s)');
            process.exit(1);
        }
        updateLambdaList.push({
            uri: uri,
            sourceDir: sourceDir,
            domain: domain,
            region: region
        });
    } else if (uri) {
        // skillSchema has been preprocessed by fn: generateSubmittingReadySkillJson
        // so that if there is uri (functionName), then the funcationName has not been found
        // from the user's AWS, so that I can directly put into the createLambdaList group.
        createLambdaList.push({
            uri: uri,
            sourceDir: sourceDir,
            domain: domain,
            region: region,
            customName: true
        });
    }
}
