// ask-cli/lib/new/new.js
'use strict';

const CONSTANTS = require('../utils/constants');
const fs = require('fs');
const inquirer = require('inquirer');
const path = require('path');

const jsonUtility = require('../utils/json-utility');
const profileHelper = require('../utils/profile-helper');
const template = require('../utils/template');
const gitTemplates = require('./git-templates');
const installNodeModule = require('../utils/install-node-module');
const skillParser = require('../utils/skill-parser');
const displayUtil = require('../utils/display-utility');

// Public
module.exports = {
    createCommand: (commander) => {
        // TODO add debug flag, installNodeModule can pass in this debug flag
        commander
            .command('new')
            .description('create a new skill project on your computer')
            .option('-n, --skill-name <name>', 'create new skill project with skill name')
            .option('-p, --profile <profile>', 'create new skill project under the chosen profile')
            .option('--lambda-name <lambda-name>', 'define lambda name if the skill needs lambda function')
            .option('--template [template-name]', 'generate a skill package locally from' +
                ' the chosen template')
            .option('--url <url>', 'the url to template list, default to Alexa Github repos')
            .option('-h, --help', 'output usage information', () => {
                console.log(CONSTANTS.COMMAND.NEW.HELP_DESCRIPTION);
                process.exit(0);
            })
            .action(handle);

        function handle(options) {
            if (typeof options === 'string') {
            console.error('[Error]: Invalid command. Please run "ask new -h" for help.');
                return;
            }

            let profile = profileHelper.runtimeProfile(options.profile);
            let url = 'https://s3.amazonaws.com/ask-cli/templates.json';
            if (options.url) {
                url = options.url;
            }

            if (options.template) {
                let templateName = (typeof options.template === 'boolean') ? null : options.template;

                gitTemplates.clone(url, options.skillName, templateName, profile, (err, skillName) => {
                    if (err) {
                        console.error(err);
                        return;
                    } else {
                        console.log("Skill '" + skillName + "' has been created based on the chosen template");
                        return;
                    }
                });
            } else {
                getSkillName(options.skillName, (skillName) => {
                    let lambdaName;
                    if (options.lambdaName) {
                        if (!skillParser.isValidLambdaFunctionName(options.lambdaName)) {
                            console.warn('Lambda function name is invalid. ' +
                                'Pattern of the name should be /[a-zA-Z0-9-_]+/.');
                            return;
                        } else {
                            lambdaName = options.lambdaName.trim();
                        }
                    } else {
                        lambdaName = skillName.replace(/ /g, '-').replace(/"/g, '');
                    }
                    newSkillScaffold(skillName, profile, lambdaName);
                });
            }
        }
    }
};

// Private
function newSkillScaffold(skillName, profile, lambdaName) {
    let currentDir = process.cwd();
    fs.access(currentDir, (fs.constants || fs).W_OK, (err) => {
        if (err) {
            console.error('No permission to write for the current path.');
            return;
        }
        let newProjDir = path.join(currentDir, skillName);
        if (fs.existsSync(newProjDir)) {
            console.warn('Failed to create skill project with the name already existed.');
            return;
        }

        let configFilePath = path.join(newProjDir, '.ask', 'config');
        let modelFilePath = path.join(newProjDir, 'models', 'en-US.json');
        let skillFilePath = path.join(newProjDir, 'skill.json');

        fs.mkdirSync(newProjDir);
        fs.mkdirSync(path.join(newProjDir, '.ask'));
        fs.mkdirSync(path.join(newProjDir, 'models'));

        template.copyConfig(configFilePath, '', false, profile);
        template.copyModel(modelFilePath);
        template.copySkill(skillFilePath, skillName);

        let AWSProfile = profileHelper.getAWSProfile(profile);
        if (AWSProfile && AWSProfile.length !== 0) {
            let defaultEndpoint = {
                "endpoint": {
                    "sourceDir": "lambda/custom"
                }
            };
            let addingObjectPath = ['manifest', 'apis', 'custom'];
            jsonUtility.writeToProperty(skillFilePath, addingObjectPath, defaultEndpoint);
            newLambdaJob(newProjDir, skillFilePath, lambdaName, profile);
        } else {
            console.log('New project for Alexa skill created.');
        }
    });
}

function newLambdaJob(currentDir, skillFilePath, lambdaName, profile) {
    let installSpinner = new displayUtil.Spinner(' Initializing Alexa skill project...');
    installSpinner.start();
    fs.mkdirSync(path.join(currentDir, 'lambda'));
    fs.mkdirSync(path.join(currentDir, 'lambda', 'custom'));
    let configFilePath = path.join(currentDir, '.ask', 'config');

    let lambdaPath = path.join(currentDir, 'lambda');
    let customLambdaPath = path.join(lambdaPath, 'custom');
    template.copyLambda(customLambdaPath);
    template.insertLambda(lambdaName, configFilePath, profile);

    let manifest = jsonUtility.read(skillFilePath);
    let skillInfo = skillParser.extractSkillInfo(manifest);

    installNodeModule.install(skillInfo, currentDir, (err) => {
        installSpinner.stop();
        if (err) {
            console.log(err);
        } else {
            console.log('New project for Alexa skill created.');
        }
    });
}

function getSkillName(skillName, callback) {
    if (skillName) {
        callback(skillName.trim());
    } else {
        inquirer.prompt([{
            message: 'Please type in your new skill name:\n',
            type: 'input',
            name: 'skillName'
        }]).then((answer) => {
            let skillName = answer.skillName.trim();
            if (skillName.length === 0) {
                console.error('[Error]: Invalid skill name.');
                process.exit(1);
            }
            callback(skillName);
        });
    }
}
